### DISCUSSION TITLE
Embedding in input/output structs with omitting fields

### DISCUSSION THREAD

**Discussion by pboguslawski** [swaggest/rest#118] [Q&A] [answered]
Is it possible to omit some fields of embedded struct in input and output interactor structs?

[Solution with `omit`](http://attilaolah.eu/2014/09/10/json-and-struct-composition-in-go/) like below doesn't hide `ID` in `openapi.json` when `ItemCreateRequest` is used as interactors `input` type:

```
type omit *struct{}

type Item struct {
	ID   uint64 `json:"id" required:"true"`
	Name string `json:"name" required:"true"`
}

type ItemCreateRequest struct {
	Item
	ID omit `json:"id,omitempty"`
}
```
If omiting fields is not possible - any other best practice to avoid repeating same field definition on every struct that uses this field (i.e. in above example we try to `inherit` `Name` field definition from `Item` and avoid to repeat its definition in `ItemCreateRequest` but we don't want to allow `ItemCreateRequest` requests to contain `ID` field).

**Comment by vearutop** [comment, marked as answer]
Mentioned solution works for marshaling in some conditions, however it is possible to have a value that results in an `object` for an "omitted" field:
```go
package main

import (
	"encoding/json"
	"fmt"
)

func main() {
	type omit *struct{}

	type Item struct {
		ID   uint64 `json:"id" required:"true"`
		Name string `json:"name" required:"true"`
	}

	type ItemCreateRequest struct {
		Item
		ID omit `json:"id,omitempty"`
	}

	ic := ItemCreateRequest{}
	ic.Item.ID = 123
	ic.Item.Name = "abc"
	ic.ID = &struct{}{}

	j, _ := json.MarshalIndent(ic, "", " ")
	fmt.Println(string(j))
}
```
```json
{
 "name": "abc",
 "id": {}
}
```

It means, that static schema of such field would be `{"type":"object"}`, not absence of field. It would be possible to make special case for such constructs, but it does not feel right due to hackyness of the solution idea.

I personally prefer precise composition with field groups to hiding fields:
```go
type ItemData struct {
	Name string `json:"name" required:"true"`
}

type Item struct {
	ID   uint64 `json:"id" required:"true"`
	ItemData
}

type ItemCreateRequest = ItemData
```

**Comment by pboguslawski** [comment]
Thanks for answer.

> I personally prefer precise composition with field groups to hiding fields:

This works but was hoping to avoid "quirks" like

```
item = &Item{
	ID: 1,
	ItemData: ItemData{
        Name: "test"
    }
}
```

when creating structs with literals.
